### PR TITLE
Extend API for classroom version, switch to Swagger for API docs

### PR DESCRIPTION
**Pull request by sderickson**
Switched us to Swagger, which provides a [spec](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md) for APIs and an ecosystem of related libraries. Some possible later improvements:

* Split the API into several files, using schema references
* Have API endpoints use swagger files to parse and enforce parameters
* Document internal-use routes as well
* [Generate a JS client](https://github.com/swagger-api/swagger-js), which both server tests and client (such as Backbone models/collections) can use.

[Check out the docs](http://19nqw4-codecombat-staging-codecombat.runnableapp.com/api-docs), which is just an interface generated from `swagger.yaml`.

Also set up `PUT /api/users/:handle/license` so that third parties can grant created users access to the classroom version. As part of that, refactored some more redeem code into Prepaid.redeem.

### CODE REVIEW

File: server/middleware/auth.coffee [hunk 1]
```diff
@@ -179,7 +179,13 @@ module.exports =
     
     activity = req.user.trackActivity 'login', 1
     yield req.user.update {activity: activity}
-    res.redirect '/'
```
Comment (basicer) [inline server/middleware/auth.coffee]: Should the clever login logic also use this logic of sending the user somewhere more specific then the homepage?  Should this logic be the auth middleware redirectHome / redirectAfterLogin?

File: server/middleware/api.coffee [hunk 2]
```diff
@@ -113,11 +113,46 @@ putUserSubscription = wrap (req, res) ->
...
+    throw new errors.NotFound('User not found.')
+
+  unless req.client._id.equals(user.get('clientCreator'))
+    throw new errors.Forbidden('Must have created the user to perform this action.')
 
+  { ends } = req.body
+  unless ends and DATETIME_REGEX.test(ends)
+    throw new errors.UnprocessableEntity('ends is not a properly formatted.')
+
+  now = new Date().toISOString()
```
Comment (basicer) [inline server/middleware/api.coffee]: I would compare using Date#getTime().  I'd assume that the ISO String sort lexically, but longs are cleaner.  Also im paranoid about like Locale stuff going ary.

File: server/middleware/api.coffee [hunk 3]
```diff
@@ -113,11 +113,46 @@ putUserSubscription = wrap (req, res) ->
...
+  if ends < now
+    throw new errors.UnprocessableEntity('ends must be in the future.')
+
+  # if the user is already subscribed, this prepaid starts when it would have ended, otherwise it starts now 
+  { endDate } = user.get('coursePrepaid') ? {}
+  if endDate and endDate >= now
```
Comment (basicer) [inline server/middleware/api.coffee]: Id guess you'd have to parse endDate here to compare by unix timestamp.

File: server/middleware/auth.coffee (hunk 1 above)
Comment (sderickson) [inline server/middleware/auth.coffee]: Good thinking!

Comment (sderickson) [inline server/middleware/auth.coffee]: No login clever test :( [Added a couple](https://github.com/codecombat/codecombat/commit/54484ea4d4e4ac4601f6ac370b8332dc803820a5#diff-061bf41f4b40f427d638fcbab5a5b0d6R322) to make sure this change doesn't break clever integration.

**Pull request merged**